File: Sources/Core/AgfeoBridge.swift
import AppKit
import Foundation

/// Alles, was der `ProfileController` von der AGFEO-Seite braucht. Als
/// Protokoll, damit der Controller ohne AppKit und ohne laufendes Dashboard
/// testbar bleibt.
protocol ProfileActivating: Sendable {
    /// Schaltet das Rufprofil. Rückgabe sagt nur, ob der Befehl abgesetzt
    /// werden konnte — eine Bestätigung liefert das Dashboard nicht.
    func activate(profile name: String) async -> Bool
}

/// Schaltet Rufprofile über den Protocol Handler des AGFEO-Dashboards.
///
/// Der Handler ist eine Einbahnstraße: kein Rückkanal, keine Bestätigung, kein
/// Auslesen des aktiven Profils oder der vorhandenen Profile (SPEC §9).
/// Deshalb sind Profilnamen im UI Freitext und die App spricht immer von
/// „zuletzt gesendet", nie von „aktiv".
final class AgfeoBridge: ProfileActivating {
    /// Ermittelt mit `osascript -e 'id of app "AGFEO-Dashboard"'` (13.08.2026),
    /// gegengeprüft gegen `CFBundleURLTypes` in der Info.plist des Dashboards,
    /// das die Schemes `adashboard`, `tksuite` und `tel` registriert.
    static let dashboardBundleID = "de.agfeo.dashboard"

    /// So lange wird nach einem Kaltstart auf den Prozess gewartet.
    private let launchTimeout: TimeInterval = 15
    /// Und so lange danach noch, bevor der Befehl kommt: das Dashboard muss
    /// sich erst mit der Anlage synchronisieren, sonst geht er ins Leere.
    private let settleAfterLaunch: TimeInterval = 2
    /// Abstand des einen Wiederholversuchs, wenn `open` fehlschlägt.
    private let retryDelay: TimeInterval = 2

    // MARK: URL

    /// Baut `adashboard:activate_call_profile?name=…`.
    ///
    /// Leerzeichen und Umlaute sind in Profilnamen zulässig. Aus
    /// `.urlQueryAllowed` werden zusätzlich `&=+?#` entfernt — die sind darin
    /// erlaubt, würden aber einen Namen wie „Büro & Mobil" in zwei
    /// Query-Parameter zerlegen.
    static func makeURL(profile name: String) -> URL? {
        var allowed = CharacterSet.urlQueryAllowed
        allowed.remove(charactersIn: "&=+?#")
        guard let encoded = name.addingPercentEncoding(withAllowedCharacters: allowed),
              !encoded.isEmpty
        else { return nil }
        return URL(string: "adashboard:activate_call_profile?name=\(encoded)")
    }

    // MARK: Schalten

    func activate(profile name: String) async -> Bool {
        guard let url = Self.makeURL(profile: name) else {
            Log.error(.agfeo, "Profilname \"\(name)\" ergibt keine gültige URL")
            return false
        }

        if !Self.isDashboardRunning {
            Log.info(.agfeo, "Dashboard läuft nicht, wird gestartet")
            guard await launchDashboard() else { return false }
            // Prozess da heißt noch nicht empfangsbereit.
            try? await Task.sleep(nanoseconds: UInt64(settleAfterLaunch * 1_000_000_000))
        }

        if await open(url) {
            Log.info(.agfeo, "Profil \"\(name)\" gesendet")
            return true
        }

        try? await Task.sleep(nanoseconds: UInt64(retryDelay * 1_000_000_000))
        if await open(url) {
            Log.notice(.agfeo, "Profil \"\(name)\" erst im zweiten Versuch gesendet")
            return true
        }

        Log.error(.agfeo, "Profil \"\(name)\" konnte nicht gesendet werden")
        return false
    }

    /// Synchrone Variante für Ruhezustand, Herunterfahren und Beenden.
    ///
    /// Startet das Dashboard bewusst **nicht**: dafür bleibt keine Zeit, und
    /// ohne laufendes Dashboard lässt sich die Anlage ohnehin nicht schalten.
    func activateNow(profile name: String) -> Bool {
        guard let url = Self.makeURL(profile: name) else { return false }
        guard Self.isDashboardRunning else {
            Log.error(
                .agfeo,
                "Dashboard läuft nicht — \"\(name)\" kann jetzt nicht mehr gesendet werden",
                immediate: true)
            return false
        }
        return NSWorkspace.shared.open(url)
    }

    // MARK: Dashboard

    static var isDashboardRunning: Bool {
        !NSRunningApplication.runningApplications(withBundleIdentifier: dashboardBundleID).isEmpty
    }

    @MainActor
    private func open(_ url: URL) -> Bool {
        NSWorkspace.shared.open(url)
    }

    /// Startet das Dashboard und wartet, bis der Prozess auftaucht.
    private func launchDashboard() async -> Bool {
        guard let appURL = await MainActor.run(body: {
            NSWorkspace.shared.urlForApplication(withBundleIdentifier: Self.dashboardBundleID)
        }) else {
            Log.error(.agfeo, "AGFEO-Dashboard ist nicht installiert (\(Self.dashboardBundleID))")
            return false
        }

        let configuration = NSWorkspace.OpenConfiguration()
        configuration.activates = false
        do {
            try await NSWorkspace.shared.openApplication(at: appURL, configuration: configuration)
        } catch {
            Log.error(.agfeo, "Dashboard-Start fehlgeschlagen: \(error.localizedDescription)")
            return false
        }

        let deadline = Date().addingTimeInterval(launchTimeout)
        while Date() < deadline {
            if Self.isDashboardRunning { return true }
            try? await Task.sleep(nanoseconds: 250_000_000)
        }

        Log.error(.agfeo, "Dashboard war nach \(Int(launchTimeout)) s nicht erreichbar")
        return false
    }
}

File: windows/src/AgfeoPresenceBridge.Core/AgfeoBridge.cs
using System.Diagnostics;

namespace AgfeoPresenceBridge.Core;

/// <summary>
/// Schaltet Rufprofile über den Protocol Handler des AGFEO-Dashboards.
/// </summary>
/// <remarks>
/// Der Handler ist eine Einbahnstraße: keine Bestätigung, kein Auslesen des
/// aktiven Profils oder der vorhandenen Profile. Deshalb sind Profilnamen im
/// Programm Freitext und die Oberfläche spricht von „zuletzt gesendet“, nie von
/// „aktiv“.
///
/// Laut Handbuch kann der Handler mehr als das eine hier genutzte Kommando:
/// <code>
/// adashboard:activate_call_profile?name=&lt;NAME&gt;     genutzt
/// adashboard:toggle_do_not_disturb?value=[on|off]  Anrufschutz
/// adashboard:group_login?name=&lt;NAME&gt;              Gruppe einbuchen
/// adashboard:group_logout?name=&lt;NAME&gt;             Gruppe ausbuchen
/// adashboard:activate_dial_rule?name=&lt;NAME&gt;       Wahlregel
/// adashboard:toggle_incognito?value=[off|on]       Rufnummernübertragung
/// adashboard:accept | reject | hangup | hold | switch | transfer | conference
/// </code>
/// </remarks>
public sealed class AgfeoBridge : IProfileActivator
{
    /// <summary>Bezeichner des Dashboard-Prozesses unter Windows.</summary>
    public const string ProcessName = "ctimon";

    /// <summary>
    /// Baut <c>adashboard:activate_call_profile?name=…</c>.
    /// </summary>
    /// <remarks>
    /// Leerzeichen und Umlaute sind in Profilnamen zulässig. Zusätzlich zu den
    /// üblichen Zeichen werden <c>&amp;=+?#</c> kodiert — sie wären in einer
    /// Query erlaubt, würden aber einen Namen wie „Büro &amp; Mobil“ in zwei
    /// Parameter zerlegen.
    /// </remarks>
    public static string? BuildUrl(string profileName)
    {
        if (string.IsNullOrEmpty(profileName)) return null;

        var encoded = new System.Text.StringBuilder();
        foreach (char character in profileName)
        {
            bool safe = char.IsAsciiLetterOrDigit(character) || character is '-' or '.' or '_' or '~';
            if (safe) encoded.Append(character);
            else
            {
                foreach (byte value in System.Text.Encoding.UTF8.GetBytes(character.ToString()))
                    encoded.Append('%').Append(value.ToString("X2"));
            }
        }
        return $"adashboard:activate_call_profile?name={encoded}";
    }

    public static bool IsDashboardRunning => Process.GetProcessesByName(ProcessName).Length > 0;

    /// <summary>
    /// Synchrone Variante für das Ende der Sitzung.
    /// </summary>
    /// <remarks>
    /// Startet das Dashboard bewusst nicht: Beim Herunterfahren bleibt dafür
    /// keine Zeit, und ohne laufendes Dashboard lässt sich die Anlage ohnehin
    /// nicht schalten.
    /// </remarks>
    public bool ActivateNow(string profileName)
    {
        string? url = BuildUrl(profileName);
        if (url is null) return false;

        if (!IsDashboardRunning)
        {
            Log.Error($"Dashboard läuft nicht — \"{profileName}\" kann jetzt nicht mehr gesendet werden");
            return false;
        }

        try
        {
            using var process = Process.Start(new ProcessStartInfo
            {
                FileName = url,
                UseShellExecute = true,
                WindowStyle = ProcessWindowStyle.Minimized,
            });
            return true;
        }
        catch (Exception error)
        {
            Log.Error($"Profil \"{profileName}\" konnte nicht gesendet werden: {error.Message}");
            return false;
        }
    }

    public Task<bool> ActivateAsync(string profileName)
    {
        string? url = BuildUrl(profileName);
        if (url is null)
        {
            Log.Error($"Profilname \"{profileName}\" ergibt keine gültige URL");
            return Task.FromResult(false);
        }

        try
        {
            // UseShellExecute lässt Windows den registrierten Handler
            // aufrufen. Das Fenster des Dashboards bleibt dabei im
            // Hintergrund — der Fokus soll nicht mitten im Tippen wegspringen.
            using var process = Process.Start(new ProcessStartInfo
            {
                FileName = url,
                UseShellExecute = true,
                WindowStyle = ProcessWindowStyle.Minimized,
            });
            Log.Info($"Profil \"{profileName}\" gesendet");
            return Task.FromResult(true);
        }
        catch (Exception error)
        {
            Log.Error($"Profil \"{profileName}\" konnte nicht gesendet werden: {error.Message}");
            return Task.FromResult(false);
        }
    }
}
